ingest/bin/transform-date-fields

```diff
@@ -36,14 +42,79 @@ def format_date(date_string: str, expected_formats: set) -> str:
     >>> format_date("2020-01-15T00:00:00Z", expected_formats)
     '2020-01-15'
     """
+    # Potential directives that datetime accepts that can return the correct year, month, day fields
+    # see https://docs.python.org/3.9/library/datetime.html#strftime-and-strptime-format-codes
+    #
+    # Allows us to check if year/month/day are included in the date format so we
+    # know when to mask incomplete dates with 'XX'
+    all_field_directives = {'%c', '%x',
+        ('%G', '%V', '%A'), ('%G', '%V', '%a'), ('%G', '%V', '%w'), ('%G', '%V', '%u')
+    }
+    month_and_day_directives = {'%j',
+        ('%U', '%A'), ('%U', '%a'), ('%U', '%w'), ('%U', '%u'),
+        ('%W', '%A'), ('%W', '%a'), ('%W', '%w'), ('%W', '%u')
+    }
```

Comment:
Hmm, I'm not sure it's valid to combine ISO week-based numbering (`%G`, `%u`, `%V`) with non-ISO week-based numbering (`%U`, `%W`, `%a`, `%A`, `%w`), as the two systems are "indexed" differently. So we may want to drop `('%U', '%u')` and `('%W', '%u')` from `month_and_day_directives` and `('%G', '%V', '%A'), ('%G', '%V', '%a'), ('%G', '%V', '%w')` from `all_field_directives`.

Reply:
Hmm, I included them since they were valid combinations for `datetime.strptime` based on the error message when trying to parse an incomplete ISO directive: 
```
ValueError: ISO year directive '%G' must be used with the ISO week directive '%V' and a weekday directive ('%A', '%a', '%w', or '%u').
```

I did see that the Year/week directives cannot be interchanged between ISO/non-ISO, I think that matches what you mean by "indexed" differently? I would think the weekday directives can be used across ISO/non-ISO directives.

Reply:
Ah, hmm. I guess they're technically valid, in the sense that datetimes can be formatted with them and re-parsed back successfully. Ok to keep them then.

I guess I was more thinking about if they're valid for reasonable use, as you can't tell unambiguously from just the formatted datetime how it should be parsed, so it's a bit "dangerous" to mix them. For example, `2022 01 Sun` (year, week, weekday) means different days depending on how it was produced:

```
>>> datetime.strptime("2022 01 Sun", "%Y %U %a")
datetime.datetime(2022, 1, 2, 0, 0)
>>> datetime.strptime("2022 01 Sun", "%G %V %a")
datetime.datetime(2022, 1, 9, 0, 0)
```

By "indexed" differently, I mean the starting points for counting weeks in a year and the weekday of weeks.

Reply:
Yeah, I guess we would be leaving it up to the user to understand the datetime directives and provide the correct combinations. The same problem exists even within non-ISO:
```
>>> datetime.strptime("2022 01 Sun", "%Y %U %a")
datetime.datetime(2022, 1, 2, 0, 0)
>>> datetime.strptime("2022 01 Sun", "%Y %W %a")
datetime.datetime(2022, 1, 9, 0, 0)
```

Reply:
Yep! Ok to leave as is.